### ISSUE TITLE
docker-compose v2.1.1 exposes problem with trailing slashes in PiHole compose file examples

### ISSUE THREAD

**Issue by Paraphraser** [pi-hole/docker-pi-hole#947]
This is a: Bug  

## Details

To be clear, I am not reporting a bug in PiHole. I am reporting a problem with PiHole **documentation** which has been exposed by a change in how up-to-date versions of docker-compose process `volumes` directives in compose files. 

The example compose files for PiHole provided here on GitHub and on DockerHub include:

```
volumes:
- ./volumes/pihole/etc-pihole/:/etc/pihole/
- ./volumes/pihole/etc-dnsmasq.d/:/etc/dnsmasq.d/
```

Under version 2.1.1 of docker-compose, the trailing slashes on those path components cause an error when the left-hand-sides of those `volumes` directives do not exist when `docker-compose up -d` is invoked. The error message is:

```
Error response from daemon: invalid mount config for type "bind": bind source path does not exist:
```

Removing the trailing slashes cures the problem.

This problem was **not** apparent with 1.x versions of docker-compose. It *may* have been apparent with 2.x versions of docker-compose prior to 2.1.1 but I have not tested to determine precisely when this problem first surfaced. All I can really say for certain is that it is there **now**, is **consistent** and is **repeatable**.

Because the problem is only triggered by the combination of up-to-date docker-compose **plus** missing left-hand-sides of `volumes` directives, the people most likely to be affected are new users on first install and those trying to debug PiHole problems by erasing persistent storage. It will not affect a PiHole user who simply upgrades their version of docker-compose.

PiHole is not the only container that is going to bang into this. I found 5 problematic containers in [SensorsIot/IOTstack](https://github.com/SensorsIot/IOTstack), including PiHole. I have submitted Pull Requests on SensorsIot/IOTstack. I'm opening this issue because you might want to inform your wider audience by changing the PiHole documentation on GitHub and DockerHub to reflect this new externally-imposed reality.

## Related Issues

- [✔︎] I have searched this repository/Pi-hole forums for existing issues and pull requests that look similar 

## How to reproduce the issue 

1. Environment data

  * Operating System: Raspbian (Buster or Bullseye)
  * Hardware: Raspberry Pi 3B+, 4B (4GB) or PiZeroW2 (but probably all RPi)
  * Kernel Architecture: Both armv7l and aarch64
  * Docker Install Info and version: 
    - Software source: Official Docker "convenience script" - `curl -fsSL https://get.docker.com | sudo sh`
    - Supplementary Software: [SensorsIot/IOTstack](https://github.com/SensorsIot/IOTstack)

		```
		Client: Docker Engine - Community
		 Version:           20.10.11
		 API version:       1.41
		 Go version:        go1.16.9
		 Git commit:        dea9396
		 Built:             Thu Nov 18 00:36:48 2021
		 OS/Arch:           linux/arm
		 Context:           default
		 Experimental:      true
		
		Server: Docker Engine - Community
		 Engine:
		  Version:          20.10.11
		  API version:      1.41 (minimum version 1.12)
		  Go version:       go1.16.9
		  Git commit:       847da18
		  Built:            Thu Nov 18 00:35:05 2021
		  OS/Arch:          linux/arm
		  Experimental:     false
		 containerd:
		  Version:          1.4.12
		  GitCommit:        7b11cfaabd73bb80907dd23182b9347b4245eb5d
		 runc:
		  Version:          1.0.2
		  GitCommit:        v1.0.2-0-g52b36a2
		 docker-init:
		  Version:          0.19.0
		  GitCommit:        de40ad0
		```
  * docker-compose install info and version:

	- 1.29.2 installed via `sudo pip3 install -U docker-compose`
	- 2.1.1 installed from [github.com/docker/compose/releases](https://github.com/docker/compose/releases)  

  * Hardware architecture: ARMv7

2. docker-compose.yml contents, docker run shell command, or paste a screenshot of any UI based configuration of containers here

	* See test cases below

3. any additional info to help reproduce

	* See test cases below

## summary of test results

![PiHoleBugTable](https://user-images.githubusercontent.com/34226495/143797528-55fb223e-28af-4bfe-902b-7a3b4d16807d.png)

### compose file (WITH trailing slashes)

```
$ cat docker-compose.yml
services:

  pihole:
    container_name: pihole
    image: pihole/pihole:latest
    restart: unless-stopped
    profiles:
      - all-hosts-off
      - iot-hub
      - sec-dev
    environment:
      - TZ=Australia/Sydney
      - WEBPASSWORD=pihole
      - INTERFACE=eth0
    ports:
      - "53:53/tcp"
      - "53:53/udp"
      - "67:67/udp"
      - "8089:80/tcp"
    volumes:
      - ./volumes/pihole/etc-pihole/:/etc/pihole/
      - ./volumes/pihole/etc-dnsmasq.d/:/etc/dnsmasq.d/
    dns:
      - 127.0.0.1
      - 1.1.1.1
    cap_add:
      - NET_ADMIN
```

### test 1 - docker-compose 1.29.2 - succeeds

1. Show version of docker-compose in use:

	```
	$ docker-compose.1.29.2 version
	docker-compose version 1.29.2, build unknown
	docker-py version: 5.0.3
	CPython version: 3.9.2
	OpenSSL version: OpenSSL 1.1.1k  25 Mar 2021
	```

2. Erase persistent storage:

	```
	$ sudo rm -rf ./volumes/pihole
	$ 
	```
	
3. Start PiHole

	```
	$ docker-compose.1.29.2 up -d
	Creating network "iotstack_default" with the default driver
	Creating pihole ... done
	```

4. Show persistent storage initialised:

	```
	$ tree ./volumes/pihole
	./volumes/pihole
	├── etc-dnsmasq.d
	│   ├── 01-pihole.conf
	│   └── 06-rfc6761.conf
	└── etc-pihole
	    ├── custom.list
	    ├── dns-servers.conf
	    ├── gravity.db
	    ├── gravity_old.db
	    ├── list.1.raw.githubusercontent.com.domains
	    ├── list.1.raw.githubusercontent.com.domains.sha1
	    ├── localbranches
	    ├── local.list
	    ├── localversions
	    ├── migration_backup
	    │   └── adlists.list
	    ├── pihole-FTL.conf
	    ├── pihole-FTL.db
	    ├── setupVars.conf
	    └── setupVars.conf.update.bak
	
	3 directories, 16 files
	```

5. Stop PiHole

	```
	$ docker-compose.1.29.2 down
	Stopping pihole ... done
	Removing pihole ... done
	Removing network iotstack_default
	```

### test 2 - docker-compose 2.1.1 - fails

1. Show version of docker-compose in use:

	```
	$ docker-compose.2.1.1 version
	Docker Compose version v2.1.1
	```

2. Erase persistent storage:

	```
	$ sudo rm -rf ./volumes/pihole
	$ 
	```
	
3. Start PiHole

	```
	$ docker-compose.2.1.1 up -d
	[+] Running 1/1
	 ⠿ Network iotstack_default  Created                                                                                   0.4s
	 ⠋ Container pihole          Creating                                                                                  0.0s
	Error response from daemon: invalid mount config for type "bind": bind source path does not exist: /home/pi/IOTstack/volumes/pihole/etc-pihole
	```

4. Show persistent storage was not initialised:

	```
	$ ls -ld ./volumes/pihole
	ls: cannot access './volumes/pihole': No such file or directory
	```
	
	I think it would be more correct to say that the external paths were not created by docker-compose, therefore the connections to the internal paths could not be created, therefore the PiHole container could not be started.

### compose file (WITHOUT trailing slashes)

```
$ cat docker-compose.yml
services:

  pihole:
    container_name: pihole
    image: pihole/pihole:latest
    restart: unless-stopped
    profiles:
      - all-hosts-off
      - iot-hub
      - sec-dev
    environment:
      - TZ=Australia/Sydney
      - WEBPASSWORD=pihole
      - INTERFACE=eth0
    ports:
      - "53:53/tcp"
      - "53:53/udp"
      - "67:67/udp"
      - "8089:80/tcp"
    volumes:
      - ./volumes/pihole/etc-pihole:/etc/pihole
      - ./volumes/pihole/etc-dnsmasq.d:/etc/dnsmasq.d
    dns:
      - 127.0.0.1
      - 1.1.1.1
    cap_add:
      - NET_ADMIN
```

### test 1 - docker-compose 1.29.2 - succeeds

1. Erase persistent storage:

	```
	$ sudo rm -rf ./volumes/pihole
	$ 
	```

2. Start PiHole

	```
	$ docker-compose.1.29.2 up -d
	Creating network "iotstack_default" with the default driver
	Creating pihole ... done
	```

3. Show persistent storage initialised:

	```
	$ tree ./volumes/pihole
	./volumes/pihole
	├── etc-dnsmasq.d
	│   ├── 01-pihole.conf
	│   └── 06-rfc6761.conf
	└── etc-pihole
	    ├── custom.list
	    ├── dns-servers.conf
	    ├── gravity.db
	    ├── gravity_old.db
	    ├── list.1.raw.githubusercontent.com.domains
	    ├── list.1.raw.githubusercontent.com.domains.sha1
	    ├── localbranches
	    ├── local.list
	    ├── localversions
	    ├── migration_backup
	    │   └── adlists.list
	    ├── pihole-FTL.conf
	    ├── pihole-FTL.db
	    ├── setupVars.conf
	    └── setupVars.conf.update.bak
	
	3 directories, 16 files
	```

4. Stop PiHole

	```
	$ docker-compose.1.29.2 down
	Stopping pihole ... done
	Removing pihole ... done
	Removing network iotstack_default
	```

### test 2 - docker-compose 2.1.1 - succeeds

1. Erase persistent storage:

	```
	$ sudo rm -rf ./volumes/pihole
	$
	```
	
2. Start PiHole

	```
	$ docker-compose.2.1.1 up -d
	[+] Running 2/2
	 ⠿ Network iotstack_default  Created                                                                                   0.4s
	 ⠿ Container pihole          Started                                                                                   1.3s
	```

3. Show persistent storage initialised:

	```
	$ tree ./volumes/pihole
	./volumes/pihole
	├── etc-dnsmasq.d
	│   ├── 01-pihole.conf
	│   └── 06-rfc6761.conf
	└── etc-pihole
	    ├── custom.list
	    ├── dns-servers.conf
	    ├── gravity.db
	    ├── gravity_old.db
	    ├── list.1.raw.githubusercontent.com.domains
	    ├── list.1.raw.githubusercontent.com.domains.sha1
	    ├── localbranches
	    ├── local.list
	    ├── localversions
	    ├── migration_backup
	    │   └── adlists.list
	    ├── pihole-FTL.conf
	    ├── pihole-FTL.db
	    ├── setupVars.conf
	    └── setupVars.conf.update.bak
	
	3 directories, 16 files
	```

4. Stop PiHole

	```
	$ docker-compose.2.1.1 down
	[+] Running 2/2
	 ⠿ Container pihole          Removed                                                                                   4.2s
	 ⠿ Network iotstack_default  Removed                                                                                   0.3s
	```

### compose file (WITH trailing slashes)

```
$ cat docker-compose.yml
services:

  pihole:
    container_name: pihole
    image: pihole/pihole:latest
    restart: unless-stopped
    profiles:
      - all-hosts-off
      - iot-hub
      - sec-dev
    environment:
      - TZ=Australia/Sydney
      - WEBPASSWORD=pihole
      - INTERFACE=eth0
    ports:
      - "53:53/tcp"
      - "53:53/udp"
      - "67:67/udp"
      - "8089:80/tcp"
    volumes:
      - ./volumes/pihole/etc-pihole/:/etc/pihole/
      - ./volumes/pihole/etc-dnsmasq.d/:/etc/dnsmasq.d/
    dns:
      - 127.0.0.1
      - 1.1.1.1
    cap_add:
      - NET_ADMIN
```

### test - simulated migration - succeeds

1. Erase persistent storage:

	```
	$ sudo rm -rf ./volumes/pihole
	$ 
	```

2. 	Start PiHole with docker-compose 1.29.2

	```
	$ docker-compose.1.29.2 up -d
	Creating network "iotstack_default" with the default driver
	Creating pihole ... done
	```

3. Show persistent storage initialised:

	```
	$ tree ./volumes/pihole
	./volumes/pihole
	├── etc-dnsmasq.d
	│   ├── 01-pihole.conf
	│   └── 06-rfc6761.conf
	└── etc-pihole
	    ├── custom.list
	    ├── dns-servers.conf
	    ├── gravity.db
	    ├── gravity_old.db
	    ├── list.1.raw.githubusercontent.com.domains
	    ├── list.1.raw.githubusercontent.com.domains.sha1
	    ├── localbranches
	    ├── local.list
	    ├── localversions
	    ├── migration_backup
	    │   └── adlists.list
	    ├── pihole-FTL.conf
	    ├── pihole-FTL.db
	    ├── setupVars.conf
	    └── setupVars.conf.update.bak
	
	3 directories, 16 files
	```

4. Stop PiHole with docker-compose 1.29.2

	```
	$ docker-compose.1.29.2 down
	Stopping pihole ... done
	Removing pihole ... done
	Removing network iotstack_default
	```

5. Start PiHole with docker-compose 2.1.1 (simulated migration)

	```
	$ docker-compose.2.1.1 up -d
	[+] Running 2/2
	 ⠿ Network iotstack_default  Created                                                                                   0.3s
	 ⠿ Container pihole          Started                                                                                   1.3s
	```

6. Other output (in case it is useful):

	```
	$ docker ps
	CONTAINER ID   IMAGE                  COMMAND      CREATED          STATUS                    PORTS                                                                              NAMES
	372464b7618e   pihole/pihole:latest   "/s6-init"   55 seconds ago   Up 53 seconds (healthy)   0.0.0.0:53->53/udp, 0.0.0.0:53->53/tcp, 0.0.0.0:67->67/udp, 0.0.0.0:8089->80/tcp   pihole
	
	pi@sec-dev:~/IOTstack$ docker logs pihole
	[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
	[s6-init] ensuring user provided files have correct perms...exited 0.
	[fix-attrs.d] applying ownership & permissions fixes...
	[fix-attrs.d] 01-resolver-resolv: applying... 
	[fix-attrs.d] 01-resolver-resolv: exited 0.
	[fix-attrs.d] done.
	[cont-init.d] executing container initialization scripts...
	[cont-init.d] 20-start.sh: executing... 
	 ::: Starting docker specific checks & setup for docker pihole/pihole
	
	  [i] Installing configs from /etc/.pihole...
	  [i] Existing dnsmasq.conf found... it is not a Pi-hole file, leaving alone!
	  [✓] Installed /etc/dnsmasq.d/01-pihole.conf
	  [✓] Installed /etc/dnsmasq.d/06-rfc6761.conf
	Existing DNS servers detected in setupVars.conf. Leaving them alone
	::: Pre existing WEBPASSWORD found
	DNSMasq binding to default interface: eth0
	Added ENV to php:
				"PHP_ERROR_LOG" => "/var/log/lighttpd/error.log",
				"ServerIP" => "0.0.0.0",
				"CORS_HOSTS" => "",
				"VIRTUAL_HOST" => "0.0.0.0",
	Using IPv4 and IPv6
	::: setup_blocklists now setting default blocklists up: 
	::: TIP: Use a docker volume for /etc/pihole/adlists.list if you want to customize for first boot
	::: Blocklists (/etc/pihole/adlists.list) now set to:
	https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts
	::: Testing pihole-FTL DNS: FTL started!
	::: Testing lighttpd config: Syntax OK
	::: All config checks passed, cleared for startup ...
	::: Enabling Query Logging
	  [i] Enabling logging...
	  [✓] Logging has been enabled!
	 ::: Docker start setup complete
	  Checking if custom gravity.db is set in /etc/pihole/pihole-FTL.conf
	  Pi-hole version is v5.6 (Latest: v5.6)
	  AdminLTE version is v5.8 (Latest: v5.8)
	  FTL version is v5.11 (Latest: v5.11)
	  Container tag is: 2021.11
	[cont-init.d] 20-start.sh: exited 0.
	[cont-init.d] done.
	[services.d] starting services
	Starting lighttpd
	Starting crond
	Starting pihole-FTL (no-daemon) as root
	[services.d] done.
	Stopping pihole-FTL
	kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec]
	Starting pihole-FTL (no-daemon) as root
	 ```

## These common fixes didn't work for my issue

- ✔︎ I have tried removing/destroying my container, and re-creating a new container
- ✔︎ I have tried fresh volume data by backing up and moving/removing the old volume data
- ✘ I have tried running the stock `docker run` example(s) in the readme (removing any customizations I added)

	> not relevant - the problem is related to newer versions of docker-compose

- ✘ I have tried a newer or older version of Docker Pi-hole (depending what version the issue started in for me)

	> not relevant - the problem is related to newer versions of docker-compose

- ✔︎ I have tried running without my volume data mounts to eliminate volumes as the cause

	> I think that's implied by starting each test with a clean slate and erasing persistent storage

If the above debugging / fixes revealed any new information note it here.
Add any other debugging steps you've taken or theories on root cause that may help.

## Workaround

Creating the left-hand-sides of the `volumes` directives, by hand, before starting PiHole with docker-compose 2.1.1, avoids the problem:

```
$ sudo rm -rf ./volumes/pihole
$ sudo mkdir -p ./volumes/pihole/etc-pihole ./volumes/pihole/etc-dnsmasq.d
$ grep "./volumes" docker-compose.yml
      - ./volumes/pihole/etc-pihole/:/etc/pihole/
      - ./volumes/pihole/etc-dnsmasq.d/:/etc/dnsmasq.d/
$ docker-compose.2.1.1 up -d
[+] Running 2/2
 ⠿ Network iotstack_default  Created                                                                                                                                                                                    0.3s
 ⠿ Container pihole          Started                                                                                                                                                                                    1.3s
```

**Comment by PromoFaux** [comment]
Thanks for the very detailed report, if the issue is one of documentation - would you mind setting up a PR against the `dev` branch with suggested changes? thanks

**Comment by Paraphraser** [comment]
@PromoFaux 

> Thanks for the very detailed report, if the issue is one of documentation - would you mind setting up a PR against the `dev` branch with suggested changes? thanks

Done. PR949.

**Comment by PromoFaux** [comment]
Addressed in #949
